### ISSUE TITLE
Bug: Designated initializers for C structs broken in C translation

### ISSUE THREAD

**Issue by bahnwaerter** [ultimate-pa/ultimate#693]
Designated initializers for C structs according to C99 are broken in Ultimate in version 0.3.0-dev-f28124e. The C translation plugin does not seem to correctly handle the lookup of C struct field names from specified initializers, thus violating the assertion in the [`CTranslationUtil` in line 253](https://github.com/ultimate-pa/ultimate/blob/f28124e2b6a385359beace8bbab307ac9eac3b63/trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/CTranslationUtil.java#L253).

The following reduced program from a large embedded program can be used to reproduce the issue:
```c
typedef unsigned char __uint8_t;
typedef short unsigned int __uint16_t;
typedef long unsigned int __uint32_t;
         
typedef __uint8_t uint8_t;
typedef __uint16_t uint16_t;
typedef __uint32_t uint32_t;

typedef struct {
    uint8_t Channel;
    uint8_t Direction;
    uint16_t Value;
    uint8_t Gain;
} SB17_SENSOR_DATA_t;

typedef struct {
    uint32_t Timestamp;
    SB17_SENSOR_DATA_t Reading;
} SB17_SENSOR_READING_t;

typedef struct {
    /* ... */
    uint32_t default_value;
} cyhal_timer_t;

uint32_t cyhal_timer_read(const cyhal_timer_t *obj);

static SB17_SENSOR_DATA_t _sb17_read_single_sensor_value();

static cyhal_timer_t timestamp_timer;

int main()
{
    /* ... */
    SB17_SENSOR_DATA_t data = _sb17_read_single_sensor_value();
    /* ... */
    SB17_SENSOR_READING_t reading = { .Timestamp = cyhal_timer_read(&timestamp_timer), .Reading = data };
    /* ... */
    return 0;
}
```

**Comment by schuessf** [comment]
Thanks for your report! In principle designated initializer seem to work, the problem only seems to occur in the combination with nested struct. But I will have a look at this problem.

**Comment by schuessf** [comment]
I tried to debug it and I minimized this example:
```c
typedef struct { int x; } DATA;
typedef struct { DATA d; } RESULT;

int main() {
  DATA data;
  data.x = 42;
  RESULT result = { .d = data };
  //@ assert result.d.x == 42;
}
```

It looks like the problem is somewhere in `InitializationHandler::constructIndexToInitInfo`: There is some queue for children-initializer-results and I have the impression that something goes wrong with `peekFirst` (that keeps the first element in the queue) and `pollFirst` (that removes the first element), but I didn't succeed to fix it.

Also: We also crash with `RESULT result = { data };` as an initialization, but which a different error message (`conversion from CStruct not implemented.`), even though the underlying problem seems to be the same.

**Comment by maul-esel** [comment]
It seems the issue is in the method `TypeHandler::isCompatibleType`, which is called by `constructIndexToInitInfo`. I've pushed a partial fix (see commit message for details) that solves the issue for the examples shown here. But I'm not sure one cannot construct other examples for which `isCompatibleType` still behaves incorrectly and causes the same error to occur.

I find the C standard on this issue particularly hard to parse. Even the first sentence of the relevant section (6.2.7) is quite extraordinary ;-)
> Two types have compatible type if their types are the same.

@bahnwaerter Do you have an overview of this "compatible types" notion in C?

**Comment by danieldietsch** [comment]
I recommend reading https://en.cppreference.com/w/c/language/compatible_type.html

**Comment by maul-esel** [comment]
@danieldietsch Thanks, that is a useful summary.

If I am reading this correctly, the "compatibility" relation is symmetric and reflexive (as stated by the extraordinary sentence quoted above), but not transitive (e.g. two different enums are both compatible with `int` but not each other).

I will see if I can translate this summary into a more complete Java implementation.

**Comment by maul-esel** [comment]
I've put together an implementation [here](https://github.com/ultimate-pa/ultimate/compare/dev...wip/dk/compatible-types). Waiting for the nightly build to see if everything is ok.

**Comment by maul-esel** [comment]
After an additional change yesterday to fix the one failing test, everything looks good. Changes are merged, so I closed this issue.
